internal/rbd/rbd_healer.go

```diff
@@ -0,0 +1,202 @@
+	if !ok {
+		util.ErrorLogMsg("parse userID failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse userID")
+	}
+	userKey, ok := secret.Data["userKey"]
+	if !ok {
+		util.ErrorLogMsg("parse userKey failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse userKey")
+	}
+	encryptionPassphrase, ok := secret.Data["encryptionPassphrase"]
+	if !ok {
+		util.ErrorLogMsg("parse encryptionPassphrase failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse encryptionPassphrase")
+	}
+	deviceSecret = map[string]string{
+		"encryptionPassphrase": string(encryptionPassphrase),
+		"userID":               string(userID),
+		"userKey":              string(userKey),
+	}
+
+	return deviceSecret, nil
+}
+
+func callNodeStageVolume(ns *NodeServer, c *k8s.Clientset, pv *v1.PersistentVolume, pluginPath string) error {
+	publishContext := make(map[string]string)
+
+	volID := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+	if acquired := ns.VolumeLocks.TryAcquire(volID); !acquired {
+		util.ErrorLogMsg(util.VolumeOperationAlreadyExistsFmt, volID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volID)
+	}
+	defer ns.VolumeLocks.Release(volID)
+
+	stagingParentPath := pluginPath + "/kubernetes.io/csi/pv/" +
+		pv.Name + "/globalmount"
+
+	util.DefaultLog("sending nodeStageVolume for volID: %s, stagingPath: %s",
+		volID, stagingParentPath)
+
+	deviceSecret, err := getSecret(c,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Namespace,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Name)
+	if err != nil {
+		util.ErrorLogMsg("getSecret failed for volID: %s, err: %v", volID, err)
+		return err
+	}
+
+	volumeContext := pv.Spec.PersistentVolumeSource.CSI.VolumeAttributes
+	volumeContext["volumeHealerContext"] = "true"
+
+	req := &csi.NodeStageVolumeRequest{
+		VolumeId:          volID,
+		PublishContext:    publishContext,
+		StagingTargetPath: stagingParentPath,
+		VolumeCapability: &csi.VolumeCapability{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: accessModeStrToInt(string(pv.Spec.AccessModes[0])),
+			},
+		},
+		Secrets:       deviceSecret,
+		VolumeContext: volumeContext,
+	}
+	if pv.Spec.PersistentVolumeSource.CSI.FSType == "block" {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Block{
+			Block: &csi.VolumeCapability_BlockVolume{},
+		}
+	} else {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{
+				FsType:     pv.Spec.PersistentVolumeSource.CSI.FSType,
+				MountFlags: pv.Spec.MountOptions,
+			},
+		}
+	}
+
+	_, err = ns.NodeStageVolume(context.TODO(), req)
+	if err != nil {
+		util.ErrorLogMsg("nodeStageVolume request failed, volID: %s, stagingPath: %s, err: %v",
+			volID, stagingParentPath, err)
+		return err
+	}
+
+	return nil
+}
+
+// RunVolumeHealer heal the volumes attached on a node.
+func RunVolumeHealer(ns *NodeServer, conf *util.Config) error {
```

Comment:
Does this function need to be exposed?

Reply:
sorry, I missing something. How this can be done? can you please elaborate on this one?

Reply:
I remember one of your comments above about partial failures, and I remember adding thoughts on similar lines.

Can we give users a way to trigger the healer to address partial failure cases?
May be like:
./cephcsi --run_healer

This will again get the volume attachments and send the NodeStageVolume Call to all the volumes.
All we need to do is makesure that the NodeStageCall is Idempotent, i.e if the process is already running just return success.

But I'm not sure if we can manage to get {*NodeServer, util.Config }?

Thoughts?

Reply:
As discussed, just restarting the pod is good rather than introducing a new way.